sdk/eventhub/azure-eventhub/azure/eventhub/aio/_producer_client_async.py

```diff
@@ -257,22 +483,86 @@ def from_connection_string(
+        If `on_error` is passed while `buffered_mode` is False when instantiating the producer client,
+        instead of error being raised from the send methods,
+        the `on_error` callback will be called with the error information related to sending.
+
+        :param event_data: The `EventData` object to be sent.
+        :type event_data: Union[~azure.eventhub.EventData, ~azure.eventhub.amqp.AmqpAnnotatedMessage]
+        :keyword float timeout: The maximum wait time to send the event data.
+         If not specified, the default wait time specified when the producer was created will be used.
+        :keyword str partition_id: The specific partition ID to send to. Default is None, in which case the service
+         will assign to all partitions using round-robin.
+         A `TypeError` will be raised if partition_id is specified and event_data_batch is an `EventDataBatch` because
+         `EventDataBatch` itself has partition_id.
+        :keyword str partition_key: With the given partition_key, event data will be sent to
+         a particular partition of the Event Hub decided by the service.
+         A `TypeError` will be raised if partition_key is specified and event_data_batch is an `EventDataBatch` because
+         `EventDataBatch` itself has partition_key.
+         If both partition_id and partition_key are provided, the partition_id will take precedence.
+         **WARNING: Setting partition_key of non-string value on the events to be sent is discouraged
+         as the partition_key will be ignored by the Event Hub service and events will be assigned
+         to all partitions using round-robin. Furthermore, there are SDKs for consuming events which expect
+         partition_key to only be string type, they might fail to parse the non-string value.**
+        """
+        partition_id = kwargs.get("partition_id") or ALL_PARTITIONS
+        partition_key = kwargs.get("partition_key")
+        timeout = kwargs.get("timeout")
+        try:
+            try:
+                await cast(EventHubProducer, self._producers[partition_id]).send(
+                    event_data, partition_key=partition_key, timeout=timeout
+                )
+            except (KeyError, AttributeError, EventHubError):
+                await self._start_producer(partition_id, timeout)
+                await cast(EventHubProducer, self._producers[partition_id]).send(
+                    event_data, partition_key=partition_key, timeout=timeout
+                )
+            if self._on_success:
+                await self._on_success([event_data], partition_id)
+        except Exception as exc:  # pylint: disable=broad-except
+            if self._on_error:
+                await self._on_error([event_data], partition_id, exc)
+            else:
+                raise
+
     async def send_batch(
         self,
         event_data_batch: Union[EventDataBatch, SendEventTypes],
-        *,
-        timeout: Optional[Union[int, float]] = None,
```

Comment:
just curious about why this was removed?

Reply:
it is not removed completely -- it's being put into the kwargs
I remember Anna has talked about putting trivial key word argument into the kwargs bag such that the signature is cleaner.